extensions/amp-ad-network-doubleclick-impl/0.1/amp-ad-network-doubleclick-impl.js

```diff
@@ -481,6 +505,96 @@ export class AmpAdNetworkDoubleclickImpl extends AmpA4A {
+   * This is a way for us to implement a simple stale-while-revalidate
+   * model.
+   * @return {?Promise} The rtc targeting info to attach to the ad url.
+   * @private
+   */
+  executeRtc_() {
+    if (rtcPromise) {
+      return this.mergeRtc();
+    }
+    const ampRtcPageElement = document.getElementById('amp-rtc');
+    if (!ampRtcPageElement) {
+      return Promise.resolve();
+    }
+    let endpoint;
+    const rtcConfig = tryParseJson(ampRtcPageElement.textContent);
+    if (!isObject(rtcConfig) || !(endpoint = rtcConfig['endpoint'])
+        || typeof endpoint != 'string' || !isSecureUrl(endpoint)) {
+      user().warn(TAG, 'invalid RTC config...');
+      return Promise.resolve();
+    }
+    this.shouldSendRequestWithoutRtc = err => {
+      return rtcConfig['sendAdRequestOnFailure'] !== false ?
+          Promise.resolve() : Promise.reject(err);
+    };
+    let rtcTotalTime;
+    const disableSWR = (
+        rtcConfig['disableStaleWhileRevalidate'] === true);
+    const headers = new Headers();
+    headers.append('Cache-Control', 'max-age=0');
+    const xhrInit = {credentials: 'include'};
+    const startTime = Date.now();
+    rtcPromise = timerFor(window).timeoutPromise(
+        RTC_TIMEOUT,
+        xhrFor(this.win).fetchJson(
+            endpoint, xhrInit).then(res => {
+              rtcTotalTime = Date.now() - startTime;
+              if (!disableSWR) {
+                // Repopulate the cache.
+                xhrFor(this.win).fetchJson(endpoint, {
+                  credentials: 'include',
+                  headers,
+                }).catch(err => {
+                  user().error(err);
+                });
+              }
+              // Redirects and non-200 status codes are forbidden for RTC.
+              if (!!res.redirected || res.status != 200) {
+                return this.shouldSendRequestWithoutRtc();
+              }
+              return res.json().then(rtcResponse => {
+                return [rtcResponse, rtcTotalTime];
+              });
+            }));
+
+    return this.mergeRtc();
+  }
+
+  mergeRtc() {
+    // add reasons for promise.reject
+    return rtcPromise.then(
+        r => {
+          const rtcResponse = r[0];
+          const rtcTotalTime = r[1];
+          // TODO :: Need to add the time on to the ad request.
+          if (rtcResponse) {
+            if (!!rtcResponse['targeting'] ||
+                !!rtcResponse['categoryExclusions']) {
+              this.jsonTargeting_['targeting'] = deepMerge(
+                  this.jsonTargeting_['targeting'] || {},
+                  rtcResponse['targeting'] || {});
+              this.jsonTargeting_['categoryExclusions'] = deepMerge(
+                  this.jsonTargeting_['categoryExclusions'] || {},
+                  rtcResponse['categoryExclusions'] || {});
+            }
+            return Promise.resolve(rtcTotalTime);
+          }
+          return this.shouldSendRequestWithoutRtc();
```

Comment:
In the event of timeout, can we send a special value in the ad request so we can track how often it occurs?  Also, Kleber requested that we include the domain of the rtc request in the ad request so we can see rtc response time by common end point